tests/wlr_foreign_toplevel_management_v1.cpp

```diff
@@ -0,0 +1,740 @@
+ */
+
+#include "helpers.h"
+#include "in_process_server.h"
+#include "foreign_toplevel_management_v1.h"
+#include "xdg_shell_stable.h"
+
+#include <gmock/gmock.h>
+#include <boost/throw_exception.hpp>
+
+using namespace testing;
+
+namespace
+{
+int const w = 10, h = 15;
+
+class ForeignToplevelManagerTest
+    : public wlcs::StartedInProcessServer
+{
+    wlcs::CheckInterfaceExpected expect_foreign_supported{the_server(), zwlr_foreign_toplevel_manager_v1_interface};
+};
+
+class ForeignToplevelHandleTest
+    : public wlcs::StartedInProcessServer
+{
+public:
+    ForeignToplevelHandleTest()
+        : client{the_server()},
+          manager{client},
+          surface{client},
+          xdg_surface{client, surface},
+          xdg_toplevel{xdg_surface}
+    {
+    }
+
+    auto toplevel() -> wlcs::ForeignToplevelHandle const&
+    {
+        if (manager.toplevels().empty())
+            BOOST_THROW_EXCEPTION(std::runtime_error("Manager does not know about any toplevels"));
+        if (manager.toplevels().size() > 1u)
+            BOOST_THROW_EXCEPTION(std::runtime_error(
+                "Manager knows about " + std::to_string(manager.toplevels().size()) + " toplevels"));
+        if (manager.toplevels()[0]->is_dirty())
+            BOOST_THROW_EXCEPTION(std::runtime_error("Toplevel has pending updates"));
+        return *manager.toplevels()[0];
+    }
+
+    auto toplevel(std::string const& app_id) -> wlcs::ForeignToplevelHandle const&
+    {
+        std::experimental::optional<wlcs::ForeignToplevelHandle const*> match;
+        for (auto const& i : manager.toplevels())
+        {
+            if (i->app_id() == app_id)
+            {
+                if (match)
+                    BOOST_THROW_EXCEPTION(std::runtime_error("Multiple toplevels have the same app ID " + app_id));
+                else
+                    match = i.get();
+            }
+        }
+        if (!match)
+            BOOST_THROW_EXCEPTION(std::runtime_error("No toplevels have the app ID " + app_id));
+        if (match.value()->is_dirty())
+            BOOST_THROW_EXCEPTION(std::runtime_error("Toplevel has pending updates"));
+        return *match.value();
+    }
+
+    wlcs::CheckInterfaceExpected expect_foreign_supported{the_server(), zwlr_foreign_toplevel_manager_v1_interface};
+    wlcs::Client client;
+    wlcs::ForeignToplevelManager manager;
+    wlcs::Surface surface;
+    wlcs::XdgSurfaceStable xdg_surface;
+    wlcs::XdgToplevelStable xdg_toplevel;
+};
+}
+
+TEST_F(ForeignToplevelManagerTest, supports_zwlr_foreign_toplevel_manager_v1)
+{
+    wlcs::Client client{the_server()};
+    wlcs::ForeignToplevelManager manager{client};
+    client.roundtrip();
+}
+
+TEST_F(ForeignToplevelManagerTest, does_not_detect_toplevels_when_test_creates_none)
+{
+    wlcs::Client client{the_server()};
+    wlcs::ForeignToplevelManager manager{client};
+    client.roundtrip();
+    ASSERT_THAT(manager.toplevels().size(), Eq(0u));
```

Comment:
What I want is for test cases for a Wayland protocol to actually contain direct calls to that protocol.

Since Wayland tends to require a large amount of boilerplate, that means that I *do* want each test body to repeat a bunch of boilerplate.

I think that it's more important that tests for a Wayland protocol have obvious calls to that protocol than that boilerplate be minimised. Or, rather, I want as little boilerplate as possible while *also* having direct `wl_protocol_foo` calls.